File: TestLeetCode/Tree/KthSmallestInBST.cs
/// <summary>
/// https://leetcode.com/problems/kth-smallest-element-in-a-bst/
/// </summary>
public class KthSmallestSolution {
    int count = 0;
    // Simply doing the in-order traversal
    public int KthSmallest(TreeNode root, int k)
    {
        if (root == null) return 0;
        int res = KthSmallest(root.left, k);
        if (count == k) { return res; }
        else {
            count++;
            if (count == k) { 
                return root.val;
            }
            return KthSmallest(root.right, k);
        }
    }

    public int KthSmallestIter(TreeNode root, int k) {
        var stack = new Stack<TreeNode>();
        var current = root;
        int count = 0;

        while (current != null || stack.Count > 0) {
            // Go to the leftmost node
            while (current != null) {
                stack.Push(current);
                current = current.left;
            }
            
            // Visit the node
            current = stack.Pop();
            count++;
            
            // If we've visited k nodes, return the value of the current node
            if (count == k) {
                return current.val;
            }
            
            // Go to the right node
            current = current.right;
        }
        
        return -1; // This return statement should never be reached if k is valid
    }
}

File: python/tree/kth_smallest_in_bst.py
from python.common import TreeNode


class KthSmallest:
    def kth_smallest(self, root: TreeNode | None, k: int) -> int:
        stack, current = [], root
        while stack or current:
            while current: stack.append(current); current = current.left
            current = stack.pop(); k -= 1
            if k == 0: return current.val
            current = current.right
        raise ValueError("k exceeds tree size")

    def kth_smallest_iterative(self, root: TreeNode | None, k: int) -> int: return self.kth_smallest(root, k)
